File: app/src/main/java/hellojdh/wintercoding/DetailActivity.java
package hellojdh.wintercoding;

import android.Manifest;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import com.bumptech.glide.Glide;

import java.io.File;
import java.io.IOException;

import hellojdh.wintercoding.databinding.ActivityDetailBinding;
import hellojdh.wintercoding.net.FileDownload;
import hellojdh.wintercoding.tool.BaseActivity;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/*
 * 상세 화면
 */
public class DetailActivity extends BaseActivity {
    private static final String TAG = DetailActivity.class.getSimpleName();
    private static final String savePath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
            +File.separator+"winterCoding"+File.separator;
    ActivityDetailBinding binding;
    private static String title,url;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this,R.layout.activity_detail);
        binding.setActivity(this);

        title = getIntent().getStringExtra("title");
        url = getIntent().getStringExtra("url");
        Log.d(TAG,"Title ="+title);
        Log.d(TAG,"Url ="+url);
        binding.detailTitle.setText(title);
        Glide.with(this).load(url).into(binding.detailImageView);
    }

    public void clickToDownload(View view){
        int permission = ContextCompat.checkSelfPermission(this,Manifest.permission.WRITE_EXTERNAL_STORAGE);
        if(permission==PackageManager.PERMISSION_DENIED) {
            ActivityCompat.requestPermissions(DetailActivity.this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 0);
            return;
        }

        File dir = new File(savePath);
        Log.d(TAG,"File path ="+dir.getAbsolutePath());
        if (!dir.exists())
            dir.mkdirs();

        final File target = new File(savePath,"/"+System.currentTimeMillis()+".jpg");
        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url(url)
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {

            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                FileDownload.downFile(target,response.body(),DetailActivity.this);
                makeToast("갤러리에 저장되었습니다");
            }
        });
    }

    void makeToast(final String t){
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(DetailActivity.this,t,Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if(grantResults[0] == PackageManager.PERMISSION_GRANTED){
            makeToast("권한이 승인되었습니다.");
        }else{
            makeToast("권한을 승인시 다운로드가 가능합니다.");
        }
    }
}

File: app/src/main/java/hellojdh/wintercoding/DetailActivity.kt
package hellojdh.wintercoding

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Environment
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.util.Log
import android.widget.Toast
import com.bumptech.glide.Glide
import hellojdh.wintercoding.net.FileDownload
import hellojdh.wintercoding.tool.BaseActivity
import kotlinx.android.synthetic.main.activity_detail.*
import okhttp3.*
import java.io.File
import java.io.IOException

/*
 * 상세 화면
 */
class DetailActivity : BaseActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail)

        titleText = intent.getStringExtra("title")
        url = intent.getStringExtra("url")
        Log.d(TAG, "Title = $title")
        Log.d(TAG, "Url = $url")

        detail_title.text = titleText
        Glide.with(this).load(url).into(detail_imageView)

        save_button.setOnClickListener { clickToDownload() }
    }

    private fun clickToDownload() {
        val permission = ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
        if (permission == PackageManager.PERMISSION_DENIED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 0)
            return
        }

        val dir = File(SAVE_PATH)
        Log.d(TAG, "File path =" + dir.absolutePath)
        if (!dir.exists())
            dir.mkdirs()

        val target = File(SAVE_PATH, "/" + System.currentTimeMillis() + ".jpg")
        val client = OkHttpClient()
        val request = Request.Builder()
                .url(url!!)
                .build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {

            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                FileDownload.downFile(target, response.body(), this@DetailActivity)
                makeToast("갤러리에 저장되었습니다")
            }
        })
    }

    internal fun makeToast(t: String) {
        runOnUiThread { Toast.makeText(this, t, Toast.LENGTH_SHORT).show() }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            makeToast("권한이 승인되었습니다.")
        } else {
            makeToast("권한을 승인시 다운로드가 가능합니다.")
        }
    }

    companion object {
        private val TAG = DetailActivity::class.java.simpleName
        private val SAVE_PATH = (Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).toString()
                + File.separator + "winterCoding" + File.separator)
        private var url: String? = null
        private var titleText: String? = null
    }
}
